src/workerd/api/streams/writable.c++

```diff
@@ -369,6 +369,173 @@ private:
+      // hold onto the ArrayBuffer for longer. We need to make sure that the backing store
+      // for the ArrayBuffer remains valid.
+      auto source = KJ_ASSERT_NONNULL(jsg::BufferSource::tryAlloc(js, size));
+      auto ptr = source.asArrayPtr();
+      memcpy(ptr.begin(), buffer, size);
+      return context.awaitJs(js, getInner().write(js, source.getHandle(js)));
+    }));
+  }
+
+  kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) override {
+    auto amount = 0;
+    for (auto& piece : pieces) {
+      amount += piece.size();
+    }
+    if (amount == 0) return kj::READY_NOW;
+    return canceler.wrap(context.run([this, amount, pieces](Worker::Lock& lock) mutable {
+      jsg::Lock& js = lock;
+
+      // Sadly, we have to allocate and copy here. Our received set of buffers are only
+      // guaranteed to live until the returned promise is resolved, but the application code
+      // may hold onto the ArrayBuffer for longer. We need to make sure that the backing store
+      // for the ArrayBuffer remains valid.
+      auto source = KJ_ASSERT_NONNULL(jsg::BufferSource::tryAlloc(js, amount));
+      auto ptr = source.asArrayPtr();
+      for (auto& piece : pieces) {
+        if (piece.size() == 0) continue;
+        KJ_ASSERT(piece.size() <= ptr.size());
+        memcpy(ptr.begin(), piece.begin(), piece.size());
+        ptr = ptr.slice(piece.size());
+      }
+
+      return context.awaitJs(js, getInner().write(js, source.getHandle(js)));
+    }));
+  }
+
+  // TODO(perf): We can't properly implement tryPumpFrom(), which means that Cap'n Proto will
+  //   be unable to perform path shortening if the underlying stream turns out to be another capnp
+  //   stream. This isn't a huge deal, but might be nice to enable someday. It may require
+  //   significant refactoring of streams.
+
+  kj::Promise<void> whenWriteDisconnected() override {
+    // TODO(soon): We might be able to support this by following the writer.closed promise,
+    // which becomes resolved when the writer is used to close the stream, or rejects when
+    // the stream has errored. However, currently, we don't have an easy way to do this for
+    // a couple of reasons.
+    //
+    // 1. The Writer's getClosed() method returns a jsg::MemoizedIdentity<jsg::Promise<void>>.
+    //    jsg::MemoizedIdentity lazily converts the jsg::Promise into a v8::Promise once it
+    //    passes through the type wrapper. It does not give us any way to consistently get
+    //    at the underlying jsg::Promise<void> or the mapped v8::Promise. We would need to
+    //    capture a TypeHandler in here and convert each time to one or the other, then
+    //    attach our continuation. It's doable but a bit of a pain.
+    //
+    // 2. The contract of whenWritedDisconnected says that it indicates when the stream will
+    //    start to reject with DISCONNETED exceptions. There is no distinction with WritableStream.
```

Comment:
I think it's fine if it resolves upon the stream being broken with any permanent error. You don't actually have to distinguish DISCONNECTED errors. The doc comment is a bit over-fitted to the raw socket use case.